Compilation error on not bleeding-edge gcc

Compiling, what I believe to be the "master" development branch, with
gcc version 4.8.5 20150623 (Red Hat 4.8.5-44)
(a Linux distribution that deliberately uses older, stable, releases)

I suddenly (i.e. must be a relatively recent change) get fatal errors:

```
   C       src/syntaxtree.c => build/obj/src/syntaxtree.c.lo
In file included from src/syntaxtree.c:21:0:
src/syntaxtree.c: In function ‘SyntaxTreeDefaultCoder’:
src/lists.h:224:5: warning: ‘vararglist’ may be used uninitialized in this function [-Wmaybe-uninitialized]
     return (*Elm0ListFuncs[TNUM_OBJ(list)])(list, pos);
     ^
src/syntaxtree.c:246:10: note: ‘vararglist’ was declared here
     Obj  vararglist;
          ^
src/syntaxtree.c: At top level:
cc1: warning: unrecognized command line option "-Wno-implicit-fallthrough" [enabled by default]
cc1: warning: unrecognized command line option "-Wno-cast-function-type" [enabled by default]
   C       src/sysfiles.c => build/obj/src/sysfiles.c.lo
   C       src/sysjmp.c => build/obj/src/sysjmp.c.lo
   C       src/sysroots.c => build/obj/src/sysroots.c.lo
   C       src/sysstr.c => build/obj/src/sysstr.c.lo
   C       src/system.c => build/obj/src/system.c.lo
   C       src/tietze.c => build/obj/src/tietze.c.lo
   C       src/tracing.c => build/obj/src/tracing.c.lo
   CXX     src/trans.cc => build/obj/src/trans.cc.lo
   C       src/vars.c => build/obj/src/vars.c.lo
   C       src/vec8bit.c => build/obj/src/vec8bit.c.lo
   C       src/vecffe.c => build/obj/src/vecffe.c.lo
   C       src/vecgf2.c => build/obj/src/vecgf2.c.lo
   C       src/vector.c => build/obj/src/vector.c.lo
   C       src/weakptr.c => build/obj/src/weakptr.c.lo
   C       src/gasman.c => build/obj/src/gasman.c.lo
   C       src/sysmem.c => build/obj/src/sysmem.c.lo
   C       src/compstat.c => build/obj/src/compstat.c.lo
   C       src/compstat_empty.c => build/obj/src/compstat_empty.c.lo
    * new link flags
    * new library flags
    * new object list
   LINK    build/gap-nocomp
   GAC     lib/oper1.g => build/c_oper1.c
   C       build/c_oper1.c => build/obj/build/c_oper1.c.lo
build/c_oper1.c:2:1: error: stray ‘\33’ in program
 /* C file produced by GAC */
 ^
build/c_oper1.c:2:2: error: expected identifier or ‘(’ before ‘[’ token
 /* C file produced by GAC */
  ^
build/c_oper1.c:2:4: error: invalid suffix "h" on integer constant
 /* C file produced by GAC */
    ^
In file included from /usr/include/stdio.h:74:0,
                 from ./src/compiled.h:20,
                 from build/c_oper1.c:3:
/usr/include/libio.h:307:3: error: unknown type name ‘size_t’
   size_t __pad5;
   ^
/usr/include/libio.h:311:67: error: ‘size_t’ undeclared here (not in a function)
   char _unused2[15 * sizeof (int) - 4 * sizeof (void *) - sizeof (size_t)];
                                                                   ^
/usr/include/libio.h:339:62: error: expected declaration specifiers or ‘...’ before ‘size_t’
 typedef __ssize_t __io_read_fn (void *__cookie, char *__buf, size_t __nbytes);
                                                              ^
/usr/include/libio.h:348:6: error: expected declaration specifiers or ‘...’ before ‘size_t’
      size_t __n);
      ^
/usr/include/libio.h:470:19: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘_IO_sgetn’
 extern _IO_size_t _IO_sgetn (_IO_FILE *, void *, _IO_size_t);
                   ^
In file included from ./src/compiled.h:20:0,
                 from build/c_oper1.c:3:
/usr/include/stdio.h:319:35: error: expected declaration specifiers or ‘...’ before ‘size_t’
 extern FILE *fmemopen (void *__s, size_t __len, const char *__modes)
                                   ^
/usr/include/stdio.h:325:47: error: expected declaration specifiers or ‘...’ before ‘size_t’
 extern FILE *open_memstream (char **__bufloc, size_t *__sizeloc) __THROW __wur;
                                               ^
/usr/include/stdio.h:337:20: error: expected declaration specifiers or ‘...’ before ‘size_t’
       int __modes, size_t __n) __THROW;
                    ^
/usr/include/stdio.h:344:10: error: expected declaration specifiers or ‘...’ before ‘size_t’
          size_t __size) __THROW;
          ^
/usr/include/stdio.h:386:44: error: expected declaration specifiers or ‘...’ before ‘size_t’
 extern int snprintf (char *__restrict __s, size_t __maxlen,
                                            ^
/usr/include/stdio.h:390:45: error: expected declaration specifiers or ‘...’ before ‘size_t’
 extern int vsnprintf (char *__restrict __s, size_t __maxlen,
                                             ^
/usr/include/stdio.h:666:11: error: expected declaration specifiers or ‘...’ before ‘size_t’
           size_t *__restrict __n, int __delimiter,
           ^
/usr/include/stdio.h:669:9: error: expected declaration specifiers or ‘...’ before ‘size_t’
         size_t *__restrict __n, int __delimiter,
         ^
/usr/include/stdio.h:679:8: error: expected declaration specifiers or ‘...’ before ‘size_t’
        size_t *__restrict __n,
        ^
/usr/include/stdio.h:709:15: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘fread’
 extern size_t fread (void *__restrict __ptr, size_t __size,
               ^
/usr/include/stdio.h:715:15: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘fwrite’
 extern size_t fwrite (const void *__restrict __ptr, size_t __size,
               ^
/usr/include/stdio.h:737:15: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘fread_unlocked’
 extern size_t fread_unlocked (void *__restrict __ptr, size_t __size,
               ^
/usr/include/stdio.h:739:15: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘fwrite_unlocked’
 extern size_t fwrite_unlocked (const void *__restrict __ptr, size_t __size,
               ^
In file included from ./src/plist.h:31:0,
                 from ./src/gap_all.h:58,
                 from ./src/compiled.h:26,
                 from build/c_oper1.c:3:
/usr/include/string.h:43:8: error: expected declaration specifiers or ‘...’ before ‘size_t’
        size_t __n) __THROW __nonnull ((1, 2));
        ^
/usr/include/string.h:46:56: error: expected declaration specifiers or ‘...’ before ‘size_t’
 extern void *memmove (void *__dest, const void *__src, size_t __n)
                                                        ^
/usr/include/string.h:55:18: error: expected declaration specifiers or ‘...’ before ‘size_t’
         int __c, size_t __n)
                  ^
/usr/include/string.h:62:42: error: expected declaration specifiers or ‘...’ before ‘size_t’
 extern void *memset (void *__s, int __c, size_t __n) __THROW __nonnull ((1));
                                          ^
/usr/include/string.h:65:56: error: expected declaration specifiers or ‘...’ before ‘size_t’
 extern int memcmp (const void *__s1, const void *__s2, size_t __n)
                                                        ^
/usr/include/string.h:92:48: error: expected declaration specifiers or ‘...’ before ‘size_t’
 extern void *memchr (const void *__s, int __c, size_t __n)
                                                ^
/usr/include/string.h:129:39: error: expected declaration specifiers or ‘...’ before ‘size_t’
         const char *__restrict __src, size_t __n)
                                       ^
/usr/include/string.h:137:9: error: expected declaration specifiers or ‘...’ before ‘size_t’
         size_t __n) __THROW __nonnull ((1, 2));
         ^
/usr/include/string.h:143:57: error: expected declaration specifiers or ‘...’ before ‘size_t’
 extern int strncmp (const char *__s1, const char *__s2, size_t __n)
                                                         ^
/usr/include/string.h:150:15: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘strxfrm’
 extern size_t strxfrm (char *__restrict __dest,
               ^
In file included from ./src/plist.h:31:0,
                 from ./src/gap_all.h:58,
                 from ./src/compiled.h:26,
                 from build/c_oper1.c:3:
/usr/include/string.h:165:15: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘strxfrm_l’
 extern size_t strxfrm_l (char *__dest, const char *__src, size_t __n,
               ^
/usr/include/string.h:180:45: error: expected declaration specifiers or ‘...’ before ‘size_t’
 extern char *strndup (const char *__string, size_t __n)
                                             ^
/usr/include/string.h:281:15: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘strcspn’
 extern size_t strcspn (const char *__s, const char *__reject)
               ^
/usr/include/string.h:285:15: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘strspn’
 extern size_t strspn (const char *__s, const char *__accept)
               ^
/usr/include/string.h:395:15: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘strlen’
 extern size_t strlen (const char *__s)
               ^
/usr/include/string.h:402:15: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘strnlen’
 extern size_t strnlen (const char *__string, size_t __maxlen)
               ^
In file included from /usr/include/features.h:375:0,
                 from /usr/include/stdio.h:27,
                 from ./src/compiled.h:20,
                 from build/c_oper1.c:3:
/usr/include/string.h:423:12: error: expected declaration specifiers or ‘...’ before ‘size_t’
 extern int __REDIRECT_NTH (strerror_r,
            ^
In file included from ./src/plist.h:31:0,
                 from ./src/gap_all.h:58,
                 from ./src/compiled.h:26,
                 from build/c_oper1.c:3:
/usr/include/string.h:447:33: error: expected declaration specifiers or ‘...’ before ‘size_t’
 extern void __bzero (void *__s, size_t __n) __THROW __nonnull ((1));
                                 ^
/usr/include/string.h:451:53: error: expected declaration specifiers or ‘...’ before ‘size_t’
 extern void bcopy (const void *__src, void *__dest, size_t __n)
                                                     ^
/usr/include/string.h:455:31: error: expected declaration specifiers or ‘...’ before ‘size_t’
 extern void bzero (void *__s, size_t __n) __THROW __nonnull ((1));
                               ^
/usr/include/string.h:458:54: error: expected declaration specifiers or ‘...’ before ‘size_t’
 extern int bcmp (const void *__s1, const void *__s2, size_t __n)
                                                      ^
/usr/include/string.h:536:61: error: expected declaration specifiers or ‘...’ before ‘size_t’
 extern int strncasecmp (const char *__s1, const char *__s2, size_t __n)
                                                             ^
/usr/include/string.h:573:34: error: expected declaration specifiers or ‘...’ before ‘size_t’
    const char *__restrict __src, size_t __n)
                                  ^
/usr/include/string.h:576:39: error: expected declaration specifiers or ‘...’ before ‘size_t’
         const char *__restrict __src, size_t __n)
                                       ^
In file included from /usr/include/string.h:633:0,
                 from ./src/plist.h:31,
                 from ./src/gap_all.h:58,
                 from ./src/compiled.h:26,
                 from build/c_oper1.c:3:
/usr/include/bits/string2.h:968:24: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘__strcspn_c1’
 __STRING_INLINE size_t __strcspn_c1 (const char *__s, int __reject);
                        ^
/usr/include/bits/string2.h:970:1: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘__strcspn_c1’
 __strcspn_c1 (const char *__s, int __reject)
 ^
/usr/include/bits/string2.h:978:24: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘__strcspn_c2’
 __STRING_INLINE size_t __strcspn_c2 (const char *__s, int __reject1,
                        ^
/usr/include/bits/string2.h:981:1: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘__strcspn_c2’
 __strcspn_c2 (const char *__s, int __reject1, int __reject2)
 ^
/usr/include/bits/string2.h:990:24: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘__strcspn_c3’
 __STRING_INLINE size_t __strcspn_c3 (const char *__s, int __reject1,
                        ^
/usr/include/bits/string2.h:993:1: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘__strcspn_c3’
 __strcspn_c3 (const char *__s, int __reject1, int __reject2,
 ^
/usr/include/bits/string2.h:1044:24: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘__strspn_c1’
 __STRING_INLINE size_t __strspn_c1 (const char *__s, int __accept);
                        ^
/usr/include/bits/string2.h:1046:1: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘__strspn_c1’
 __strspn_c1 (const char *__s, int __accept)
 ^
/usr/include/bits/string2.h:1055:24: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘__strspn_c2’
 __STRING_INLINE size_t __strspn_c2 (const char *__s, int __accept1,
                        ^
/usr/include/bits/string2.h:1058:1: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘__strspn_c2’
 __strspn_c2 (const char *__s, int __accept1, int __accept2)
 ^
/usr/include/bits/string2.h:1067:24: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘__strspn_c3’
 __STRING_INLINE size_t __strspn_c3 (const char *__s, int __accept1,
                        ^
/usr/include/bits/string2.h:1070:1: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘__strspn_c3’
 __strspn_c3 (const char *__s, int __accept1, int __accept2, int __accept3)
 ^
/usr/include/bits/string2.h: In function ‘__strpbrk_c2’:
/usr/include/bits/string2.h:1128:50: error: expected ‘;’ before ‘__s’
   return *__s == '\0' ? NULL : (char *) (size_t) __s;
                                                  ^
/usr/include/bits/string2.h: In function ‘__strpbrk_c3’:
/usr/include/bits/string2.h:1140:50: error: expected ‘;’ before ‘__s’
   return *__s == '\0' ? NULL : (char *) (size_t) __s;
                                                  ^
In file included from /usr/include/bits/string2.h:1296:0,
                 from /usr/include/string.h:633,
                 from ./src/plist.h:31,
                 from ./src/gap_all.h:58,
                 from ./src/compiled.h:26,
                 from build/c_oper1.c:3:
/usr/include/stdlib.h: At top level:
/usr/include/stdlib.h:465:22: error: expected declaration specifiers or ‘...’ before ‘size_t’
 extern void *malloc (size_t __size) __THROW __attribute_malloc__ __wur;
                      ^
/usr/include/stdlib.h:467:22: error: expected declaration specifiers or ‘...’ before ‘size_t’
 extern void *calloc (size_t __nmemb, size_t __size)
                      ^
/usr/include/stdlib.h:467:38: error: expected declaration specifiers or ‘...’ before ‘size_t’
 extern void *calloc (size_t __nmemb, size_t __size)
                                      ^
In file included from /usr/include/string.h:633:0,
                 from ./src/plist.h:31,
                 from ./src/gap_all.h:58,
                 from ./src/compiled.h:26,
                 from build/c_oper1.c:3:
/usr/include/bits/string2.h:1320:47: error: expected declaration specifiers or ‘...’ before ‘size_t’
 extern char *__strndup (const char *__string, size_t __n)
                                               ^
In file included from ./src/gap_all.h:58:0,
                 from ./src/compiled.h:26,
                 from build/c_oper1.c:3:
./src/plist.h: In function ‘NewPlistFromArray’:
./src/plist.h:326:5: warning: implicit declaration of function ‘memcpy’ [-Wimplicit-function-declaration]
     memcpy(BASE_PTR_PLIST(o), list, length * sizeof(Obj));
     ^
./src/plist.h:326:5: warning: incompatible implicit declaration of built-in function ‘memcpy’ [enabled by default]
In file included from ./src/gap_all.h:71:0,
                 from ./src/compiled.h:26,
                 from build/c_oper1.c:3:
./src/stringobj.h: In function ‘COPY_CHARS’:
./src/stringobj.h:239:5: warning: incompatible implicit declaration of built-in function ‘memcpy’ [enabled by default]
     memcpy(CHARS_STRING(str), pnt, n);
     ^
./src/stringobj.h: At top level:
./src/stringobj.h:334:55: error: expected declaration specifiers or ‘...’ before ‘size_t’
 EXPORT_INLINE Obj MakeStringWithLen(const char * buf, size_t len)
                                                       ^
./src/stringobj.h: In function ‘MakeString’:
./src/stringobj.h:343:5: warning: implicit declaration of function ‘MakeStringWithLen’ [-Wimplicit-function-declaration]
     return MakeStringWithLen(cstr, strlen(cstr));
     ^
./src/stringobj.h:343:5: warning: nested extern declaration of ‘MakeStringWithLen’ [-Wnested-externs]
./src/stringobj.h:343:5: warning: implicit declaration of function ‘strlen’ [-Wimplicit-function-declaration]
./src/stringobj.h:343:36: warning: incompatible implicit declaration of built-in function ‘strlen’ [enabled by default]
     return MakeStringWithLen(cstr, strlen(cstr));
                                    ^
./src/stringobj.h:343:5: warning: return makes pointer from integer without a cast [enabled by default]
     return MakeStringWithLen(cstr, strlen(cstr));
     ^
./src/stringobj.h: At top level:
./src/stringobj.h:353:58: error: expected declaration specifiers or ‘...’ before ‘size_t’
 EXPORT_INLINE Obj MakeImmStringWithLen(const char * buf, size_t len)
                                                          ^
In file included from ./src/gap_all.h:72:0,
                 from ./src/compiled.h:26,
                 from build/c_oper1.c:3:
./src/sysfiles.h:237:33: error: expected declaration specifiers or ‘...’ before ‘size_t’
 Int SyRead(Int fid, void * ptr, size_t len);
                                 ^
./src/sysfiles.h:238:40: error: expected declaration specifiers or ‘...’ before ‘size_t’
 Int SyWrite(Int fid, const void * ptr, size_t len);
                                        ^
./src/sysfiles.h:240:43: error: expected declaration specifiers or ‘...’ before ‘size_t’
 Int SyReadWithBuffer(Int fid, void * ptr, size_t len);
                                           ^
In file included from ./src/gap_all.h:75:0,
                 from ./src/compiled.h:26,
                 from build/c_oper1.c:3:
./src/trycatch.h: In function ‘gap_restore_trycatch’:
./src/trycatch.h:105:5: warning: incompatible implicit declaration of built-in function ‘memcpy’ [enabled by default]
     memcpy(STATE(ReadJmpError), jb, sizeof(jmp_buf));
     ^
build/c_oper1.c: At top level:
cc1: warning: unrecognized command line option "-Wno-implicit-fallthrough" [enabled by default]
cc1: warning: unrecognized command line option "-Wno-cast-function-type" [enabled by default]
make: *** [build/obj/build/c_oper1.c.lo] Error 1
rm build/c_oper1.c
```

Thank you @fingolfin and @ChrisJefferson .  I get:
```
Unix69 /home/fac/f/hulpke/tmp/gap > md5sum build/c_*
d27501209c21b3c6456e9bacf49c0934  build/c_oper1.c
```
(Yes, just one file). As for the content, the second line starts garbled:
```
Unix68 /home/fac/f/hulpke/tmp/gap > od -ha build/c_oper1.c | head
0000000    6923    6e66    6564    2066    5641    494f    5f44    5250
          #   i   f   n   d   e   f  sp   A   V   O   I   D   _   P   R
0000020    4345    4d4f    4950    454c    0a44    5b1b    313f    3330
          E   C   O   M   P   I   L   E   D  nl esc   [   ?   1   0   3
0000040    6834    2a2f    4320    6620    6c69    2065    7270    646f
          4   h   /   *  sp   C  sp   f   i   l   e  sp   p   r   o   d
0000060    6375    6465    6220    2079    4147    2043    2f2a    230a
          u   c   e   d  sp   b   y  sp   G   A   C  sp   *   /  nl   #
```
If I edit the file (thanks for not deleting it any longer) it compiles fine (and the same hold for c_type.c) So it is really just the garbage in the second line (and I have a work-around for myself).

After the patch by @fingolfin I get the same error and the second line still contains the garbage characters:
```
   C       build/c_oper1.c => build/obj/build/c_oper1.c.lo
build/c_oper1.c:1:1: error: stray ‘\33’ in program
 /* C file produced by GAC */
 ^
build/c_oper1.c:1:2: error: expected identifier or ‘(’ before ‘[’ token
 /* C file produced by GAC */
  ^
build/c_oper1.c:1:4: error: invalid suffix "h" on integer constant
 /* C file produced by GAC */
    ^
```
I attach the strace file
[out.txt](https://github.com/gap-system/gap/files/6460396/out.txt)

Thank you!

--- a/cnf/gac.in
+++ b/cnf/gac.in
@@ -357,20 +357,16 @@ make_compstat () {
         echo     "/* made by 'gac', can be thrown away */"
         echo     "#include \"compiled.h\""
 
-        echo     "#ifndef AVOID_PRECOMPILED"
         echo     "extern StructInitInfo * Init__type1 ( void );"
         echo     "extern StructInitInfo * Init__oper1( void );"
-        echo     "#endif"
 
         for name in ${names}; do
             echo "extern StructInitInfo * Init__${name} ( void );"
         done
 
         echo     "InitInfoFunc CompInitFuncs[] = {"
-        echo     "#ifndef AVOID_PRECOMPILED"
         echo     "    Init__type1,"
         echo     "    Init__oper1,"
-        echo     "#endif"
 
         for name in ${names}; do
             echo "    Init__${name},"

--- a/cnf/gap-c-gen.sh
+++ b/cnf/gap-c-gen.sh
@@ -17,12 +17,10 @@ then
 fi
 
 # invoke GAP in compiler mode
-echo "#ifndef AVOID_PRECOMPILED" > "$C_FILE.tmp"
-"$GAPBIN" -C "*stdout*" \
+"$GAPBIN" -E -C "$C_FILE.tmp" \
           "$GAP_FILE" \
           "Init_$STEM" \
           GAPROOT/lib/$STEM.g >> "$C_FILE.tmp"
-echo "#endif" >> "$C_FILE.tmp"
 
 # if the new file differs from the old one, overwrite
 if cmp -s "$C_FILE.tmp" "$C_FILE"

--- a/src/compstat.c
+++ b/src/compstat.c
@@ -11,24 +11,17 @@
 #include "compstat.h"
 #include "common.h"
 
-// #define AVOID_PRECOMPILED
-
-
 /****************************************************************************
 **
 *V  CompInitFuncs . . . . . . . . . .  list of compiled module init functions
 **
 **  This a dummy list in case no module is statically linked.
 */
-#ifndef AVOID_PRECOMPILED
 extern StructInitInfo * Init__type1(void);
 extern StructInitInfo * Init__oper1(void);
-#endif
 
 InitInfoFunc CompInitFuncs[] = {
-#ifndef AVOID_PRECOMPILED
     Init__type1,
     Init__oper1,
-#endif
     0
 };

